rq/worker.py

```diff
@@ -1573,3 +1584,208 @@ class RandomWorker(Worker):
+        """Checks whether the thread pool is full.
+        Returns True if there are no idle threads, False otherwise
+
+        Returns:
+            is_full (bool): True if full, False otherwise.
+        """
+        if self._idle_threads == 0:
+            return True
+        return False
+
+    @property
+    def default_pool_size(self) -> int:
+        """THe default TheadPool size.
+        By default, each CPU core should run N Threads,
+        where N is the `DEFAULT_CPU_THREADS``
+
+        Returns:
+            cpus (int): Number of CPUs
+        """
+        from multiprocessing import cpu_count
+        return cpu_count() * DEFAULT_CPU_THREADS
+
+    def work(
+        self,
+        burst: bool = False,
+        logging_level: str = "INFO",
+        date_format=DEFAULT_LOGGING_DATE_FORMAT,
+        log_format=DEFAULT_LOGGING_FORMAT,
+        max_jobs=None,
+        with_scheduler: bool = False,
+    ):
+        """Starts the work loop.
+
+        Pops jobs from the current list of queues, and submits each job to the ThreadPool.
+        When all queues are empty, block and wait for new jobs to arrive on any of the
+        queues, unless `burst` mode is enabled.
+
+        The return value indicates whether any jobs were processed.
+        """
+        self.bootstrap(logging_level, date_format, log_format)
+        completed_jobs = 0
+        self.log.info("ThreadPoolWorker %s: started with %s threads, version %s", self.key, self.threadpool_size, VERSION)
+        self.log.warning("*** WARNING: ThreadPoolWorker is in beta and may be unstable. Don't use it in production!")
+        if with_scheduler:
+            self._start_scheduler(burst, logging_level, date_format, log_format)
+
+        self._install_signal_handlers()
+        try:
+            while True:
+                try:
+                    self.check_for_suspension(burst)
+
+                    if self.should_run_maintenance_tasks:
+                        self.run_maintenance_tasks()
+
+                    if self._stop_requested:
+                        self.log.info('Worker %s: stopping on request', self.key)
+                        break
+
+                    if self.is_pool_full:
+                        self.log.debug('ThreadPool is full, waiting for idle threads...')
+                        self.__wait_for_slot()
+
+                    timeout = None if burst else self.dequeue_timeout
+                    result = self.dequeue_job_and_maintain_ttl(timeout)
+                    if result is None:
+                        if not burst:
+                            break
+                        has_pending_dependents = self.__check_pending_dependents()
+                        if has_pending_dependents:
+                            continue
+                        self.log.info("Worker %s: done, quitting", self.key)
+                        break
+
+                    job, queue = result
+                    self.reorder_queues(reference_queue=queue)
```

Comment:
`reorder_queues()` should also be moved into `dequeue_job_and_maintain_ttl()`, so it can adjust the order of queues before dequeueing.

Reply:
@selwin what do you think of opening a brand new PR only for the worker refactoring?
the refactor of the current `Worker` methods being a conditioning factor for a new worker implementation, it seems that it will be more efficient to work on one thing at a time (refactoring it first), otherwise we will be constantly dealing with conflicts and rebasing.

Reply:
> reorder_queues() should also be moved into dequeue_job_and_maintain_ttl(), so it can adjust the order of queues before dequeueing.

`roundrobin` uses a `reference_queue` which comes from `dequeue_job_and_maintain_ttl`

Reply:
We should be able to do this:
```python
if not reference_queue:
    reference_queue = last_queue_in_the_list  # In case of round robin, the shuffling will return the first queue
```